hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/configuration/OptionsResolver.java

```diff
@@ -480,6 +480,13 @@ public static boolean isLazyFailedWritesCleanPolicy(Configuration conf) {
     return HoodieCleanConfig.FAILED_WRITES_CLEANER_POLICY.defaultValue().equalsIgnoreCase(HoodieFailedWritesCleaningPolicy.LAZY.name());
   }
 
+  /**
+   * Returns whether the writers should use blocking instant time generation.
+   */
+  public static boolean isBlockingInstantGeneration(Configuration conf) {
+    return isCowTable(conf) && isUpsertOperation(conf);
```

Comment:
Why do we need to fall back to synchronous instant generation during upsert operations on Cow tables?

Reply:
Writer pipeline for cow with upsert: rowdataToHoodie -> bucket_assign -> writer
* bucket assign function use state to assign record location.
* writer use merge handle to upsert/merge records into the assigned file group.

If eager flush / flush triggered by checkpoint happens before previous instant committed successfully, there are two potential problems:
* if the file group has never been written, exception happens: "FileID xxx of partition path xxx does not exist."
* if the file group has base file with smaller instant, data loss may happen because the flushed records will merge with base file with wrong version.

Reply:
Got it, thanks for your explanation